src/Core/Auth/UserFeatures/DeviceTrust/UntrustDevicesCommand.cs

```diff
@@ -0,0 +1,36 @@
+﻿using Bit.Core.Entities;
+using Bit.Core.Repositories;
+
+namespace Bit.Core.Auth.UserFeatures.DeviceTrust;
+
+public class UntrustDevicesCommand : IUntrustDevicesCommand
+{
+    private readonly IDeviceRepository _deviceRepository;
+
+    public UntrustDevicesCommand(
+        IDeviceRepository deviceRepository)
+    {
+        _deviceRepository = deviceRepository;
+    }
+
+    public async Task UntrustDevices(User user, IEnumerable<Guid> devicesToUntrust)
+    {
+        var userDevices = await _deviceRepository.GetManyByUserIdAsync(user.Id);
+        foreach (var deviceId in devicesToUntrust)
+        {
+            if (!userDevices.Any(d => d.Id == deviceId))
+            {
+                throw new UnauthorizedAccessException($"User {user.Id} does not have access to device {deviceId}");
+            }
+        }
+
+        foreach (var deviceId in devicesToUntrust)
+        {
+            var device = userDevices.First(d => d.Id == deviceId);
+            device.EncryptedPrivateKey = null;
+            device.EncryptedPublicKey = null;
+            device.EncryptedUserKey = null;
+            await _deviceRepository.UpsertAsync(device);
+        }
+    }
```

Comment:
Both of the operators Any and First are O(n) in terms of time complexity. If we use a dict here, we can optimize this to achieve O(1) lookups across the board: 
```suggestion
        var userDevices = await _deviceRepository.GetManyByUserIdAsync(user.Id);
        var deviceIdDict = userDevices.ToDictionary(device => device.Id);
        
        // Validate that the user owns all devices that they passed in
        foreach (var deviceId in devicesToUntrust)
        {
            if (!deviceIdDict.ContainsKey(deviceId))
            {
                throw new UnauthorizedAccessException($"User {user.Id} does not have access to device {deviceId}");
            }
        }

        foreach (var deviceId in devicesToUntrust)
        {
            var device = deviceIdDict[deviceId];
            device.EncryptedPrivateKey = null;
            device.EncryptedPublicKey = null;
            device.EncryptedUserKey = null;
            await _deviceRepository.UpsertAsync(device);
        }
    }
```

Reply:
I ran this code in your branch and made sure the tests passed as well.